src/utils/format_json/open_telemetry.c

```diff
@@ -0,0 +1,278 @@
+
+#include <yajl/yajl_gen.h>
+
+#define CHECK(f)                                                               \
+  do {                                                                         \
+    int status = (f);                                                          \
+    if (status != 0) {                                                         \
+      ERROR("format_json: %s failed with status %d", #f, status);              \
+      return status;                                                           \
+    }                                                                          \
+  } while (0)
+
+static int json_add_string(yajl_gen g, char const *str) /* {{{ */
+{
+  if (str == NULL) {
+    CHECK(yajl_gen_null(g));
+    return 0;
+  }
+
+  CHECK(yajl_gen_string(g, (const unsigned char *)str,
+                        (unsigned int)strlen(str)));
+  return 0;
+} /* }}} int json_add_string */
+
+static int key_value(yajl_gen g, label_pair_t label) {
+  CHECK(yajl_gen_map_open(g)); /* BEGIN KeyValue */
+
+  CHECK(json_add_string(g, "key"));
+  CHECK(json_add_string(g, label.name));
+
+  CHECK(json_add_string(g, "value"));
+  CHECK(yajl_gen_map_open(g)); /* BEGIN AnyValue */
+  CHECK(json_add_string(g, "stringValue"));
+  CHECK(json_add_string(g, label.value));
+  CHECK(yajl_gen_map_close(g)); /* END AnyValue */
+
+  CHECK(yajl_gen_map_close(g)); /* END KeyValue */
+  return 0;
+}
+
+static int number_data_point(yajl_gen g, metric_t const *m) {
+  CHECK(yajl_gen_map_open(g)); /* BEGIN NumberDataPoint */
+
+  CHECK(json_add_string(g, "attributes"));
+  for (size_t i = 0; i < m->label.num; i++) {
+    CHECK(key_value(g, m->label.ptr[i]));
+  }
+  CHECK(yajl_gen_array_close(g));
+
+  CHECK(json_add_string(g, "timeUnixNano"));
+  CHECK(yajl_gen_integer(g, CDTIME_T_TO_NS(m->time)));
+
+  switch (m->family->type) {
+  case METRIC_TYPE_COUNTER:
+    CHECK(json_add_string(g, "asInt"));
+    CHECK(yajl_gen_integer(g, m->value.counter));
+    break;
+  case METRIC_TYPE_GAUGE:
+    CHECK(json_add_string(g, "asDouble"));
+    CHECK(yajl_gen_integer(g, m->value.gauge));
+    break;
+  case METRIC_TYPE_UNTYPED:
+    // TODO
+    assert(0);
+  }
+
+  CHECK(yajl_gen_map_close(g)); /* END NumberDataPoint */
+  return 0;
+}
+
+static int gauge(yajl_gen g, metric_family_t const *fam) {
+  CHECK(yajl_gen_map_open(g)); /* BEGIN Gauge */
+
+  CHECK(json_add_string(g, "dataPoints"));
+  CHECK(yajl_gen_array_open(g));
+  for (size_t i = 0; i < fam->metric.num; i++) {
+    CHECK(number_data_point(g, fam->metric.ptr + i));
+  }
+  CHECK(yajl_gen_array_close(g));
+
+  CHECK(yajl_gen_map_close(g)); /* END Gauge */
+  return 0;
+}
+
+static int sum(yajl_gen g, metric_family_t const *fam) {
+  CHECK(yajl_gen_map_open(g)); /* BEGIN Sum */
+
+  CHECK(json_add_string(g, "dataPoints"));
+  CHECK(yajl_gen_array_open(g));
+  for (size_t i = 0; i < fam->metric.num; i++) {
+    CHECK(number_data_point(g, fam->metric.ptr + i));
```

Comment:
`sum()` caller does it with `CHECK()`, so same error could be logged up to 5 times.

Reply:
Fixed in 1bd6d8c7220fe9f00e9155f4a8dc4570b200ba08